File: tests/conformance/ipc/directory_chain.py
"""Bootstrap through Rust ServiceDirectory, then call its negotiated service."""

from __future__ import annotations

import asyncio
import os
import sys
import tempfile
import time
from pathlib import Path

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "sdk" / "python"))
sys.path.insert(0, str(ROOT / "gen" / "python"))

from nlos.sabi.v1.envelope_pb2 import (  # noqa: E402
    RETRY_DIRECTIVE_DO_NOT_RETRY,
    RETRY_DIRECTIVE_QUERY_OPERATION_OR_RETRY_SAME_IDEMPOTENCY_KEY,
    SABI_ERROR_CODE_CONFLICT,
    SABI_ERROR_CODE_UNCERTAIN,
    ExchangeRequest,
)
from nlos_sdk import (  # noqa: E402
    IpcError,
    LocalRpcClient,
    MethodSemantics,
    ServiceDirectoryClient,
    ServiceRequirement,
    TransportConfig,
    validate_response_context,
)


def endpoint(label: str) -> str:
    unique = f"{os.getpid():x}-{time.time_ns() & 0xFFFFFFF:x}-{label[0]}"
    if sys.platform == "win32":
        return rf"\\.\pipe\nlos-directory-{unique}"
    return str(Path(tempfile.gettempdir()) / f"nlos-directory-{unique}.sock")


async def start_server(
    directory_endpoint: str,
    business_endpoint: str,
    authority_path: str,
    phase: str,
) -> asyncio.subprocess.Process:
    process = await asyncio.create_subprocess_exec(
        "cargo",
        "run",
        "--quiet",
        "-p",
        "nlos-ipc",
        "--features",
        "conformance-server",
        "--bin",
        "nlos-directory-chain",
        "--",
        directory_endpoint,
        business_endpoint,
        authority_path,
        phase,
        cwd=ROOT,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    assert process.stdout is not None
    ready = await asyncio.wait_for(process.stdout.readline(), timeout=60)
    if ready.strip() != b"READY":
        assert process.stderr is not None
        stderr = await process.stderr.read()
        raise AssertionError(f"directory chain failed before ready: {stderr!r}")
    return process


def business_request(
    request_seed: int,
    correlation_seed: int,
    method: str = "cancel",
    key_seed: int = 6,
    payload: bytes = b"\x04\x05\x06",
) -> ExchangeRequest:
    request = ExchangeRequest()
    request.envelope.schema.name = "nlos.sabi.Envelope"
    request.envelope.schema.major = 1
    request.envelope.schema.minor = 1
    request.envelope.request_id = bytes([request_seed]) * 16
    request.envelope.service = "operation"
    request.envelope.method = method
    request_context = request.envelope.request_context
    request_context.caller.principal_id = bytes([1]) * 16
    request_context.caller.application_id = bytes([2]) * 16
    request_context.caller.process_id = bytes([3]) * 16
    request_context.caller.process_generation = 7
    request_context.correlation_id = bytes([correlation_seed]) * 16
    request_context.idempotency_key = bytes([key_seed]) * 16
    request_context.deadline_monotonic_ns = 123_456
    capability = request_context.capability_handles.add()
    capability.slot = 11
    capability.generation = 2
    request.envelope.payload = payload
    return request


async def main() -> None:
    directory_endpoint = endpoint("bootstrap")
    business_endpoint = endpoint("business")
    authority_path = str(
        Path(tempfile.gettempdir())
        / f"nlos-directory-{os.getpid()}-{time.time_ns()}-authority.sqlite3"
    )
    server = await start_server(
        directory_endpoint,
        business_endpoint,
        authority_path,
        "commit",
    )
    try:
        transport_config = TransportConfig(
            connect_timeout=2,
            read_timeout=2,
            write_timeout=2,
        )
        connected = await ServiceDirectoryClient.negotiate_and_connect(
            directory_endpoint,
            ServiceRequirement(
                service="operation",
                schema_name="nlos.sabi.Envelope",
                major=1,
                minimum_minor=1,
            ),
            transport_config,
        )
        assert connected.binding.endpoint.address == business_endpoint
        assert connected.binding.candidate.generation == 7

        try:
            await connected.client.exchange(business_request(9, 5))
        except IpcError as error:
            assert error.code == "READ"
        else:
            raise AssertionError("first committed exchange must disconnect")
        await connected.client.close()

        assert await server.wait() == 0
        server = await start_server(
            directory_endpoint,
            business_endpoint,
            authority_path,
            "recover",
        )
        recovered = await ServiceDirectoryClient.negotiate_and_connect(
            directory_endpoint,
            ServiceRequirement(
                service="operation",
                schema_name="nlos.sabi.Envelope",
                major=1,
                minimum_minor=1,
            ),
            transport_config,
        )
        assert recovered.binding.endpoint.address == business_endpoint

        retry_client = recovered.client
        response = await retry_client.exchange(business_request(10, 7))
        assert response.envelope.request_id == bytes([10]) * 16
        assert response.envelope.payload == b"\x04\x05\x06\xd0"
        response_context = validate_response_context(
            response.envelope,
            MethodSemantics(side_effecting=True, long_running=True),
        )
        assert response_context.correlation_id == bytes([7]) * 16
        assert response_context.operation.generation == 1
        assert response_context.receipts[0].receipt_id == bytes([0x99]) * 16
        await retry_client.close()

        conflict_client = await LocalRpcClient.connect(
            business_endpoint,
            transport_config,
        )
        conflict = await conflict_client.exchange(
            business_request(11, 8, "cancel", 6, b"\x04\x05\x07")
        )
        conflict_context = validate_response_context(
            conflict.envelope,
            MethodSemantics(side_effecting=True, long_running=True),
        )
        assert conflict_context.failure.code == SABI_ERROR_CODE_CONFLICT
        assert conflict_context.failure.retry == RETRY_DIRECTIVE_DO_NOT_RETRY
        assert (
            conflict_context.operation.operation_id
            == response_context.operation.operation_id
        )
        await conflict_client.close()

        pending_client = await LocalRpcClient.connect(
            business_endpoint,
            transport_config,
        )
        pending = await pending_client.exchange(
            business_request(12, 9, "pending", 7, b"\x01")
        )
        pending_context = validate_response_context(
            pending.envelope,
            MethodSemantics(side_effecting=True, long_running=True),
        )
        assert pending_context.failure.code == SABI_ERROR_CODE_UNCERTAIN
        assert (
            pending_context.failure.retry
            == RETRY_DIRECTIVE_QUERY_OPERATION_OR_RETRY_SAME_IDEMPOTENCY_KEY
        )
        assert pending_context.HasField("operation")
        await pending_client.close()

        assert await server.wait() == 0
    except BaseException:
        server.kill()
        await server.wait()
        raise


if __name__ == "__main__":
    asyncio.run(main())

File: tests/conformance/ipc/directory_chain.ts
import assert from "node:assert/strict";
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import { tmpdir } from "node:os";
import { join } from "node:path";

import { create } from "@bufbuild/protobuf";

import {
  CallerIdentitySchema,
  CapabilityHandleSchema,
  EnvelopeSchema,
  ExchangeRequestSchema,
  RetryDirective,
  SabiRequestContextSchema,
  SabiErrorCode,
  SchemaIdentitySchema,
} from "../../../gen/typescript/nlos/sabi/v1/envelope_pb.ts";
import { validateResponseContext } from "../../../sdk/typescript/src/common.ts";
import {
  IpcError,
  LocalRpcClient,
} from "../../../sdk/typescript/src/local_rpc.ts";
import { ServiceDirectoryClient } from "../../../sdk/typescript/src/service_directory.ts";

function endpoint(label: string): string {
  const unique = `${process.pid}-${Date.now()}-${label}`;
  return process.platform === "win32"
    ? `\\\\.\\pipe\\nlos-directory-${unique}`
    : join(tmpdir(), `nlos-directory-${unique}.sock`);
}

async function startServer(
  directoryEndpoint: string,
  businessEndpoint: string,
  authorityPath: string,
  phase: "commit" | "recover",
): Promise<ChildProcessWithoutNullStreams> {
  const server = spawn(
    "cargo",
    [
      "run",
      "--quiet",
      "-p",
      "nlos-ipc",
      "--features",
      "conformance-server",
      "--bin",
      "nlos-directory-chain",
      "--",
      directoryEndpoint,
      businessEndpoint,
      authorityPath,
      phase,
    ],
    { stdio: ["pipe", "pipe", "pipe"] },
  );
  await new Promise<void>((resolve, reject) => {
    const timer = setTimeout(
      () => reject(new Error("directory chain server did not become ready")),
      30_000,
    );
    let stdout = "";
    let stderr = "";
    server.stdout.on("data", (chunk: Buffer) => {
      stdout += chunk.toString("utf8");
      if (stdout.includes("READY\n") || stdout.includes("READY\r\n")) {
        clearTimeout(timer);
        resolve();
      }
    });
    server.stderr.on("data", (chunk: Buffer) => {
      stderr += chunk.toString("utf8");
    });
    server.once("exit", (code) => {
      clearTimeout(timer);
      reject(new Error(`directory chain exited early (${code}): ${stderr}`));
    });
    server.once("error", reject);
  });
  return server;
}

function waitForExit(
  server: ChildProcessWithoutNullStreams,
): Promise<number | null> {
  if (server.exitCode !== null || server.signalCode !== null) {
    return Promise.resolve(server.exitCode);
  }
  return new Promise((resolve, reject) => {
    server.once("exit", resolve);
    server.once("error", reject);
  });
}

function businessRequest(
  requestSeed: number,
  correlationSeed: number,
  method = "cancel",
  keySeed = 6,
  payload: readonly number[] = [4, 5, 6],
) {
  return create(ExchangeRequestSchema, {
    envelope: create(EnvelopeSchema, {
      schema: create(SchemaIdentitySchema, {
        name: "nlos.sabi.Envelope",
        major: 1,
        minor: 1,
      }),
      requestId: new Uint8Array(16).fill(requestSeed),
      service: "operation",
      method,
      commonContext: {
        case: "requestContext",
        value: create(SabiRequestContextSchema, {
          caller: create(CallerIdentitySchema, {
            principalId: new Uint8Array(16).fill(1),
            applicationId: new Uint8Array(16).fill(2),
            processId: new Uint8Array(16).fill(3),
            processGeneration: 7n,
          }),
          correlationId: new Uint8Array(16).fill(correlationSeed),
          idempotencyKey: new Uint8Array(16).fill(keySeed),
          deadlineMonotonicNs: 123_456n,
          capabilityHandles: [
            create(CapabilityHandleSchema, { slot: 11n, generation: 2n }),
          ],
        }),
      },
      payload: new Uint8Array(payload),
    }),
  });
}

const directoryEndpoint = endpoint("bootstrap");
const businessEndpoint = endpoint("business");
const authorityPath = join(
  tmpdir(),
  `nlos-directory-${process.pid}-${Date.now()}-authority.sqlite3`,
);
let server = await startServer(
  directoryEndpoint,
  businessEndpoint,
  authorityPath,
  "commit",
);
try {
  const transportConfig = {
    connectTimeoutMs: 2_000,
    readTimeoutMs: 2_000,
    writeTimeoutMs: 2_000,
  };
  const connected = await ServiceDirectoryClient.negotiateAndConnect(
    directoryEndpoint,
    {
      service: "operation",
      schemaName: "nlos.sabi.Envelope",
      major: 1,
      minimumMinor: 1,
    },
    transportConfig,
  );
  assert.equal(connected.binding.endpoint?.address, businessEndpoint);
  assert.equal(connected.binding.candidate?.generation, 7n);

  await assert.rejects(
    connected.client.exchange(businessRequest(9, 5)),
    (error: unknown) => error instanceof IpcError && error.code === "READ",
  );
  connected.client.close();

  const commitCode = await waitForExit(server);
  assert.equal(commitCode, 0);

  server = await startServer(
    directoryEndpoint,
    businessEndpoint,
    authorityPath,
    "recover",
  );
  const recovered = await ServiceDirectoryClient.negotiateAndConnect(
    directoryEndpoint,
    {
      service: "operation",
      schemaName: "nlos.sabi.Envelope",
      major: 1,
      minimumMinor: 1,
    },
    transportConfig,
  );
  assert.equal(recovered.binding.endpoint?.address, businessEndpoint);

  const retryClient = recovered.client;
  const response = await retryClient.exchange(businessRequest(10, 7));
  assert.deepEqual(
    Uint8Array.from(response.envelope?.requestId ?? []),
    new Uint8Array(16).fill(10),
  );
  assert.deepEqual(
    Uint8Array.from(response.envelope?.payload ?? []),
    new Uint8Array([4, 5, 6, 0xd0]),
  );
  assert.ok(response.envelope);
  const responseContext = validateResponseContext(response.envelope, {
    sideEffecting: true,
    longRunning: true,
  });
  assert.deepEqual(
    Uint8Array.from(responseContext.correlationId),
    new Uint8Array(16).fill(7),
  );
  assert.equal(responseContext.operation?.generation, 1n);
  assert.deepEqual(
    Uint8Array.from(responseContext.receipts[0]?.receiptId ?? []),
    new Uint8Array(16).fill(0x99),
  );
  retryClient.close();

  const conflictClient = await LocalRpcClient.connect(
    businessEndpoint,
    transportConfig,
  );
  const conflict = await conflictClient.exchange(
    businessRequest(11, 8, "cancel", 6, [4, 5, 7]),
  );
  assert.ok(conflict.envelope);
  const conflictContext = validateResponseContext(conflict.envelope, {
    sideEffecting: true,
    longRunning: true,
  });
  assert.equal(conflictContext.failure?.code, SabiErrorCode.CONFLICT);
  assert.equal(conflictContext.failure?.retry, RetryDirective.DO_NOT_RETRY);
  assert.deepEqual(
    Uint8Array.from(conflictContext.operation?.operationId ?? []),
    Uint8Array.from(responseContext.operation?.operationId ?? []),
  );
  conflictClient.close();

  const pendingClient = await LocalRpcClient.connect(
    businessEndpoint,
    transportConfig,
  );
  const pending = await pendingClient.exchange(
    businessRequest(12, 9, "pending", 7, [1]),
  );
  assert.ok(pending.envelope);
  const pendingContext = validateResponseContext(pending.envelope, {
    sideEffecting: true,
    longRunning: true,
  });
  assert.equal(pendingContext.failure?.code, SabiErrorCode.UNCERTAIN);
  assert.equal(
    pendingContext.failure?.retry,
    RetryDirective.QUERY_OPERATION_OR_RETRY_SAME_IDEMPOTENCY_KEY,
  );
  assert.ok(pendingContext.operation);
  pendingClient.close();

  const code = await waitForExit(server);
  assert.equal(code, 0);
} catch (error) {
  server.kill();
  throw error;
}
